src/Common/Crypto/X509Certificate.h

```diff
@@ -0,0 +1,333 @@
+#pragma once
+
+#include <Common/Exception.h>
+
+#include <boost/container/flat_set.hpp>
+
+#include "config.h"
+
+#if USE_SSL
+#    include <openssl/bio.h>
+#    include <openssl/crypto.h>
+#    include <openssl/err.h>
+#    include <openssl/pem.h>
+#    include <openssl/x509.h>
+#    include <openssl/x509v3.h>
+#    include <Common/OpenSSLHelpers.h>
+#endif
+
+#include <memory>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int OPENSSL_ERROR;
+    extern const int BAD_ARGUMENTS;
+}
+
+
+#if USE_SSL
+class X509Certificate
+{
+public:
+    using List = std::vector<X509Certificate>;
+    constexpr static size_t NAME_BUFFER_SIZE = 256;
+
+    explicit X509Certificate(X509 * cert_)
+        : certificate(cert_)
+    {
+    }
+    explicit operator X509 *() const { return certificate; }
+
+    X509Certificate & operator=(X509Certificate && other) noexcept
+    {
+        if (this == &other)
+            return *this;
+
+        if (certificate)
+            X509_free(certificate);
+
+        certificate = other.certificate;
+        other.certificate = nullptr;
+
+        return *this;
+    }
+
+    X509Certificate(const X509Certificate &) = delete;
+    X509Certificate & operator=(const X509Certificate &) = delete;
+
+    X509Certificate(X509Certificate && other) noexcept
+        : certificate(other.certificate)
+    {
+        other.certificate = nullptr;
+    }
+
+    explicit X509Certificate(const std::string & path)
+    {
+        using BIO_ptr = std::unique_ptr<BIO, decltype(&BIO_free)>;
+        BIO_ptr bio(BIO_new(BIO_s_mem()), BIO_free);
+
+        if (!bio)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "BIO_new failed: {}", getOpenSSLErrors());
+
+        BIO * file = BIO_new_file(path.c_str(), "r");
+
+        if (!file)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "BIO_new_file failed: {}", getOpenSSLErrors());
+
+        certificate = PEM_read_bio_X509(file, nullptr, nullptr, nullptr);
+        if (!certificate)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "PEM_read_bio_X509 failed for file {}: {}", path, getOpenSSLErrors());
+    }
+
+    static X509Certificate::List fromFile(const std::string & path)
+    {
+        using BIO_ptr = std::unique_ptr<BIO, decltype(&BIO_free)>;
```

Comment:
`BIO_ptr` is a good idea. Could make use of this in KeyPair.h as well, e.g. in `fromFile`.